persistence/src/main/scala/com/mesosphere/usi/repository/TemplateRepository.scala

```diff
@@ -0,0 +1,204 @@
+import com.mesosphere.usi.repository.TemplateRepositoryLike.Template
+import com.mesosphere.usi.storage.PathTrie
+import com.mesosphere.usi.state.PathId
+import org.apache.zookeeper.KeeperException.NoNodeException
+
+import scala.async.Async.{async, await}
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.language.reflectiveCalls
+import scala.util.{Failure, Success, Try}
+
+/**
+  * This class implements a repository for templates. It uses the underlying [[ZooKeeperPersistenceStore]] and stores
+  * [[Template]]s using its [[PathId]] to determine the storage location. The absolute Zookeeper path, built from the
+  * [[base]], service's pathId and service's hash (see [[TemplateRepositoryLike.storePath]] method for details).
+  *
+  * Upon the initialization this class reads the contents of the store on it's initialization keeping everything in
+  * memory using a [[PathTrie]]. This allows for synchronous reading of the templates without blocking the thread.
+  * Thus all the reading/synchronous methods are served directly from the memory, all the writing calls e.g. [[create]],
+  * [[delete]] are changing the underlying storage first and the in-memory copy afterwards.
+  *
+  * IMPORTANT NOTE:
+  *
+  * An interesting fact about Zookeeper: one can create a lot of znodes underneath one parent but if you try to get all of
+  * them by calling [[com.mesosphere.usi.storage.zookeeper.ZooKeeperPersistenceStore.children()]] (on the parent znode)
+  * you are likely to get an error like:
+  * `
+  * java.io.IOException: Packet len20800020 is out of range!
+  *  at org.apache.zookeeper.ClientCnxnSocket.readLength(ClientCnxnSocket.java:112)
+  *  ...
+  * ```
+  *
+  * Turns out that ZK has a packet length limit which is 4096 * 1024 bytes by default:
+  * https://github.com/apache/zookeeper/blob/0cb4011dac7ec28637426cafd98b4f8f299ef61d/src/java/main/org/apache/zookeeper/client/ZKClientConfig.java#L58
+  *
+  * It can be altered by setting `jute.maxbuffer` environment variable. Packet in this context is an application level
+  * packet containing all the children names. Some experimentation showed that each child element in that packet has ~4bytes
+  * overhead for encoding. So, let's say each child znode *name* e.g. holding Marathon app definition is 50 characters long
+  * (seems like a good guess given 3-4 levels of nesting e.g. `eng_dev_databases_my-favourite-mysql-instance`), we could only
+  * have: 4096 * 1024 / (50 + 4) =  ~75k children nodes until we hit the exception.
+  *
+  * In this class we implicitly rely on the users to *not* put too many children (apps/pods) under one parent. Since that
+  * number can be quite high (~75k) I think we are fine without implementing any guards.
+  */
+class TemplateRepository(val store: ZooKeeperPersistenceStore, val base: String)(implicit val mat: Materializer)
```

Comment:
I'd drop it for now and concentrate on `ZooKeeperPersistenceStore`

Reply:
Done